### PR TITLE
[REFACTOR][RPC][PROCOTOL-CHANGE] Modularize the RPC infra

### PR DESCRIPTION
**Pull request by tqchen**
This PR refactors the RPC protocol to make it more modularized.

- RPCSession: represent a set of features that need to be implemented
- RPCEndPont: End point that forwards the RPCSession requests over a communication channel.
- RPCModule: Exposes an RPCSession as an rpc device in the TVM Runtime API.

In the new design, the local machine is presented as a special case of RPCSession.
The remote is just another client session that calls into RPCEndPoint.
The RPC communication path is as follows.

```
client -> ClientSession -> EndPoint[client@n0]
-> networking[between n0 <=> n1]
-> EndPoint[server@n1] -> LocalSession[@n1]
```

Because of the new modular design, we can now chain more sessions together.
For example, we can now run the following proxy setup (testcase in test_runtime_rpc.test_session_constructor).

```
client -> ClientSession -> Endpoint[client@n0]
-> networking[between n0 <=> n1]
-> Endpoint[server@n1] -> ClientSession -> Endpoint[client@n1]
-> networking[between n1 <=> n2]
-> Endpoint[server@n2] -> LocalSession[@n2]
```

We can also implement other types of Sessions.
For example, we could make uTVM session a special case of the RPCSession
and use the same mechanism for session management.

We also add more comments about the internal of the RPC.
The communication protocol is simplfied using a similar convention as PackedFunc.
This allows us to further reduce the amount of special remote syscalls.

Due to the major improvement and simplification, we are making a non-compatible update to the RPC protocol. It means that the client and server needs to be upgraded to together in order for it to function correctly.

This PR also introduces a versioning mechanism to the current RPC procotol,
so that future upgrade will be produce more user friendly with error messages.

### CODE REVIEW

**Comment by tqchen** [comment]
RFC https://discuss.tvm.ai/t/modularize-rpc-infra/6544

**Comment by tqchen** [comment]
cc @FrozenGene @merrymercy @eqy @ZihengJiang @areusch @ajtulloch @antinucleon @junrushao1994 @jroesch @kazum @yzhliu

File: src/runtime/rpc/rpc_channel.cc [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace tvm {
+namespace runtime {
+
+size_t CallbackChannel::Send(const void* data, size_t size) {
+  TVMByteArray bytes;
+  bytes.data = static_cast<const char*>(data);
+  bytes.size = size;
+  int64_t n = fsend_(bytes);
+  if (n == -1) {
```
Comment (areusch) [inline src/runtime/rpc/rpc_channel.cc:34]: where is this error code documented?

Comment (tqchen) [reply src/runtime/rpc/rpc_channel.cc:34]: good point, fsend is a user provided function(since it is a callback), so we will need to document it in the channel definition itself.

**Comment by tqchen** [comment]
Update: added a minrpc implementation and example for using popen to create a rpc

File: apps/cpp_rpc/rpc_server.cc [hunk 2]
```diff
@@ -217,10 +217,10 @@ class RPCServer {
    * \param opts Parsed options for socket
    * \param ping_period Timeout for select call waiting
    */
-  void AcceptConnection(TrackerClient* tracker, 
+  void AcceptConnection(TrackerClient* tracker,
```
Comment (tmoreau89) [inline apps/cpp_rpc/rpc_server.cc:220]: if spaces are ending up in the source code, should we turn on the linter

Comment (tqchen) [reply apps/cpp_rpc/rpc_server.cc:220]: this is because apps are not covered by the linter as oppose the `src/`

File: jvm/core/src/main/java/org/apache/tvm/contrib/GraphRuntime.java [hunk 3]
```diff
@@ -38,53 +38,14 @@
...
-      Function fcreate = Function.getFunction("tvm.graph_runtime.create");
-      if (fcreate == null) {
-        throw new RuntimeException("Cannot find global function tvm.graph_runtime.create."
-            + "Did you compile tvm_runtime with correct version?");
-      }
-      graphModule = fcreate.pushArg(graphJson)
-          .pushArg(libmod).pushArg(ctx.deviceType).pushArg(ctx.deviceId)
-          .invoke().asModule();
+    Function fcreate = Function.getFunction("tvm.graph_runtime.create");
+    if (fcreate == null) {
```
Comment (areusch) [inline jvm/core/src/main/java/org/apache/tvm/contrib/GraphRuntime.java:42]: would be better to add this to a new helper function `Function.getFunctionOrThrow()`. doesn't need to be in this PR though

File: python/tvm/_ffi/_ctypes/packed_func.py [hunk 4]
```diff
@@ -141,6 +141,17 @@ def _make_tvm_args(args, temp_args):
         elif isinstance(arg, TVMContext):
             values[i].v_int64 = _ctx_to_int64(arg)
             type_codes[i] = TypeCode.TVM_CONTEXT
+        elif isinstance(arg, bytes):
```
Comment (areusch) [inline python/tvm/_ffi/_ctypes/packed_func.py:144]: do you need the separate elif block here? seems like you could just make the following block `elif isinstance(arg, (bytearray, bytes))` and collapse the logic there.

File: python/tvm/_ffi/_cython/packed_func.pxi [hunk 5]
```diff
@@ -142,6 +142,18 @@ cdef inline int make_arg(object arg,
         value[0].v_ctx = (<DLContext*>(
             <unsigned long long>ctypes.addressof(arg)))[0]
         tcode[0] = kTVMContext
+    elif isinstance(arg, bytes):
```
Comment (areusch) [inline python/tvm/_ffi/_cython/packed_func.pxi:145]: same question here

File: python/tvm/_ffi/base.py [hunk 6]
```diff
@@ -48,8 +48,11 @@ def _load_lib():
     """Load libary by searching possible path."""
     lib_path = libinfo.find_lib_path()
     lib = ctypes.CDLL(lib_path[0], ctypes.RTLD_GLOBAL)
-    # DMatrix functions
     lib.TVMGetLastError.restype = ctypes.c_char_p
+    # Put the libpath to LD_LIBRARY_PATH
+    # will be useful for pipe session to find libtvm
+    os.environ["LD_LIBRARY_PATH"] = "%s:%s" % (
```
Comment (areusch) [inline python/tvm/_ffi/base.py:54]: is this just to make minrpcserver work? if so, could we just add in the search path in that binary rather than changing LD_LIBRARY_PATH for all library loads here? or at least, can you revert the change when you're done?

i think this means that all hooks that invoke a subprocess inherit the modified LD_LIBRARY_PATH

File: python/tvm/contrib/cc.py [hunk 7]
```diff
@@ -90,7 +90,8 @@ def get_target_triple():
 def cross_compiler(compile_func,
                    options=None,
                    output_format=None,
-                   get_target_triple=None):
+                   get_target_triple=None,
+                   add_files=None):
```
Comment (areusch) [inline python/tvm/contrib/cc.py:94]: this is a little bit of a weird place (compiler configuration time) to add extra files (I.e. invocation configuration time). why can't this be pushed to the invocation of fcompile?

File: python/tvm/error.py [hunk 8]
```diff
@@ -57,6 +57,11 @@ def __init__(self, msg):
 register_error("KeyError", KeyError)
 
 
+@register_error
+class RPCError(RuntimeError):
+    """Error thrown by the RPC call."""
```
Comment (areusch) [inline python/tvm/error.py:62]: can you specify more (I.e. by the transport, by the serializer/deserializer, by the remote server, or by the remote function)?

File: python/tvm/rpc/client.py [hunk 9]
```diff
@@ -183,28 +198,37 @@ class LocalSession(RPCSession):
...
+def _popen_session(binary):
+    temp = util.tempdir()
 
-    def load_module(self, path):
-        return _load_module(self._temp.relpath(path))
+    if isinstance(binary, (bytes, bytearray)):
+        path_exec = temp.relpath("server.minrpc")
+        with open(path_exec, "wb") as outfile:
+            outfile.write(binary)
+        os.chmod(path_exec, stat.S_IXUSR)
```
Comment (areusch) [inline python/tvm/rpc/client.py:213]: TIL you can run a file you can't read. but, maybe we should make it readable by the user too in case of e.g. segfault or whatnot?

File: python/tvm/rpc/client.py [hunk 10]
```diff
@@ -183,28 +198,37 @@ class LocalSession(RPCSession):
...
-        return bytearray(open(self._temp.relpath(path), "rb").read())
+@tvm._ffi.register_func("rpc.PopenSession")
+def _popen_session(binary):
+    temp = util.tempdir()
 
-    def load_module(self, path):
-        return _load_module(self._temp.relpath(path))
+    if isinstance(binary, (bytes, bytearray)):
+        path_exec = temp.relpath("server.minrpc")
+        with open(path_exec, "wb") as outfile:
```
Comment (areusch) [inline python/tvm/rpc/client.py:211]: maybe we can log the temporary file name being created here at debug level?

File: python/tvm/rpc/client.py [hunk 11]
```diff
@@ -183,28 +198,37 @@ class LocalSession(RPCSession):
...
+            outfile.write(binary)
+        os.chmod(path_exec, stat.S_IXUSR)
+        path_exec = os.path.abspath(path_exec)
+    else:
+        path_exec = os.path.abspath(binary)
```
Comment (areusch) [inline python/tvm/rpc/client.py:216]: maybe we could at least do a existence check and check it is an executable file here, since otherwise i'm not sure what error the user might see from C? also, is abspath necessary?

File: python/tvm/rpc/client.py [hunk 12]
```diff
@@ -378,7 +402,7 @@ def request_and_run(self,
                 key, max_retry, str(last_err)))
 
 
-def connect(url, port, key="", session_timeout=0):
+def connect(url, port, key="", session_timeout=0, session_constructor=None):
```
Comment (areusch) [inline python/tvm/rpc/client.py:405]: maybe we can call it session_constructor_args so it's more clear it should be a list rather than a callable?

File: python/tvm/rpc/client.py [hunk 13]
```diff
@@ -397,15 +421,41 @@ def connect(url, port, key="", session_timeout=0):
         the connection when duration is longer than this value.
         When duration is zero, it means the request must always be kept alive.
 
+    session_constructor: List
+        List of additional arguments to passed as the remote session constructor.
```
Comment (areusch) [inline python/tvm/rpc/client.py:425]: well it's more like a list, the first arg is a string specifying the name of the constructor, and following args are the positional args to that function. update docstring?

File: python/tvm/rpc/minrpc.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    server : str
+        The kind of built in minrpc server.
+
+    Returns
+    -------
+    path : str
+        The path to the min server library.
+    """
+    curr_dir = os.path.dirname(os.path.realpath(os.path.expanduser(__file__)))
+    source_dir = os.path.abspath(os.path.join(curr_dir, "..", "..", ".."))
```
Comment (areusch) [inline python/tvm/rpc/minrpc.py:37]: is there a util function we can use for this?

File: python/tvm/_ffi/_ctypes/packed_func.py (hunk 4 above)
Comment (tqchen) [reply python/tvm/_ffi/_ctypes/packed_func.py:144]: Good pt `(ctypes.char * n).from_buffer` only takes bytearray, so in the case of bytes, we have to construct a bytearray and push to the temp. 

In the case of bytearray we can directly use it as it is. We can certainly try to unify a bit by having a wrapper function to further check for bytes. That was the original reason for not unifying the two. I will try to take anoter stab

File: python/tvm/_ffi/base.py (hunk 6 above)
Comment (tqchen) [reply python/tvm/_ffi/base.py:54]: Given that it makes sense to make libtvm available to subprocesses, I think we can leave it as it is. While setting rpath for the binary would work if the binary is on the same machine, it may not work after we send it through an RPC, so it might be helpful to keep the env variable.

File: python/tvm/contrib/cc.py (hunk 7 above)
Comment (tqchen) [reply python/tvm/contrib/cc.py:94]: Indeed it is an API tradeoff. The main advantage of functor decorator style is that it is composable.

For example, we could pass use the following to attach minserver to the ndk compiler. 

```c++
mod.export_library("xyz", tvm.rpc.with_minrpc(cc.ndk))
```

Another reason is that in the case of minrpc, we need to attach a special property(need_system_lib) to tell the export_library to check for the `--system-lib` option.  

We might need a bit more thoughts to make the alternative API(as follows) work.
```c++
mod.export_library("xyz", cc.ndk, libs=[tvm.rpc.minrpc()])
```

Finally,  such functor decorator is easier to work with in the AutoTVM(we just have to pass the decorated version as build_func, instead of passing an additional argument)

Comment (tqchen) [reply python/tvm/contrib/cc.py:94]: I am not very attached to the argument though. Pehaps changig to something like `default_libs` or `add_libs` would alleviate the concern?

File: python/tvm/rpc/minrpc.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+
+def with_minrpc(compile_func,
+                server="posix_popen_server",
+                runtime="libtvm"):
+    """Attach the compiler function with minrpc related options.
+
+    Parameters
+    ----------
+    compile_func : function
```
Comment (areusch) [inline python/tvm/rpc/minrpc.py:55]: Union[str, Callable[[str, str, Optional[str]], None]] right?

File: python/tvm/contrib/cc.py (hunk 7 above)
Comment (areusch) [reply python/tvm/contrib/cc.py:94]: it seems like based on how this is used now, the problem is that we are trying to make export_library build a binary in a few small cases (but, which will become increasingly common as µTVM is more integrated). i think it's reasonable for export_library to control the compiler invocation up to the point of generating a .o. beyond that ldflags and additional statically-linked dependencies (which will have their own cflags and compiler toolchains) seem like a bit much of an API abuse. maybe this will be better addressed by the compilation and linking work i'm doing? if so, is there an intermediate hack we can do that avoids adding to this API (if you agree that's the right direction to go)?

File: python/tvm/rpc/server.py [hunk 16]
```diff
@@ -56,7 +57,7 @@ def _server_env(load_library, work_path=None):
         temp = util.tempdir()
 
     # pylint: disable=unused-variable
-    @tvm._ffi.register_func("tvm.rpc.server.workpath")
+    @tvm._ffi.register_func("tvm.rpc.server.workpath", override=True)
```
Comment (areusch) [inline python/tvm/rpc/server.py:60]: shall we update `override=True` to `can_override=True`, since it's changing in the C++?

File: src/runtime/rpc/minrpc/minrpc_server.h [hunk 17]
```diff
@@ -0,0 +1,598 @@
...
+        ret_tcode[1] = kTVMDLTensorHandle;
+        ret_value[2].v_handle = ret_value[1].v_handle;
+        ret_tcode[2] = kTVMOpaqueHandle;
+        this->ReturnPackedSeq(ret_value, ret_tcode, 3);
+      } else if (rv_tcode == kTVMPackedFuncHandle ||
+                 rv_tcode == kTVMModuleHandle) {
+        ret_tcode[1] = kTVMOpaqueHandle;
+        this->ReturnPackedSeq(ret_value, ret_tcode, 2);
+      } else {
+        this->ReturnPackedSeq(ret_value, ret_tcode, 2);
```
Comment (areusch) [inline src/runtime/rpc/minrpc/minrpc_server.h:158]: is this num_args == 1?

Comment (areusch) [reply src/runtime/rpc/minrpc/minrpc_server.h:158]: never mind

File: apps/bundle_deploy/runtime.cc [hunk 18]
```diff
@@ -16,7 +16,6 @@
  * specific language governing permissions and limitations
  * under the License.
  */
-
```
Comment (areusch) [inline apps/bundle_deploy/runtime.cc:19]: could revert this

File: apps/cpp_rpc/rpc_tracker_client.h [hunk 19]
```diff
@@ -31,7 +31,7 @@
 #include <vector>
 #include <string>
 
-#include "../../src/runtime/rpc/rpc_session.h"
+#include "../../src/runtime/rpc/rpc_end_point.h"
```
Comment (areusch) [inline apps/cpp_rpc/rpc_tracker_client.h:34]: rpc_endpoint.h

File: python/tvm/_ffi/base.py (hunk 6 above)
Comment (areusch) [reply python/tvm/_ffi/base.py:54]: i'm worried that something else might be in the same directory as libtvm.so, which is where the subtle bug may come from. I don't think it's a good idea to muck with LD_LIBRARY_PATH unless you're being really explicit about this.

File: python/tvm/rpc/server.py (hunk 16 above)
Comment (tqchen) [reply python/tvm/rpc/server.py:60]: In this case, we changed the c++ argument because override is a keyword in c++11, the python API change would be bigger and let us keep it as it is

File: src/runtime/rpc/minrpc/minrpc_server.h [hunk 20]
```diff
@@ -0,0 +1,598 @@
...
+  void ServerLoop() {
+    RPCCode code;
+    uint64_t packet_len;
+
+    while (true) {
+      arena_.RecycleAll();
+      allow_clean_shutdown_ = true;
+
+      this->Read(&packet_len);
+      if (packet_len == 0) continue;
```
Comment (areusch) [inline src/runtime/rpc/minrpc/minrpc_server.h:84]: is there a max we should also test against?

File: src/runtime/rpc/minrpc/minrpc_server.h [hunk 21]
```diff
@@ -0,0 +1,598 @@
...
+
+    this->Read(&handle);
+    this->Read(&offset);
+    this->Read(&num_bytes);
+    this->Read(&ctx);
+    this->Read(&type_hint);
+
+    char* data_ptr;
+    int call_ecode = 0;
+    if (ctx.device_type == kDLCPU) {
```
Comment (areusch) [inline src/runtime/rpc/minrpc/minrpc_server.h:178]: I think here we don't need this check anymore, right? we always want the else case? should we add a CHECK?

File: src/runtime/rpc/minrpc/minrpc_server.h [hunk 22]
```diff
@@ -0,0 +1,598 @@
...
+      data_ptr = reinterpret_cast<char*>(handle) + offset;
+    } else {
+      data_ptr = this->ArenaAlloc<char>(num_bytes);
```
Comment (areusch) [inline src/runtime/rpc/minrpc/minrpc_server.h:181]: CHECK or return an error if this fails?

Comment (areusch) [reply src/runtime/rpc/minrpc/minrpc_server.h:181]: er, we should return an error and the reason i'm raising this is i'm wondering which one

Comment (areusch) [reply src/runtime/rpc/minrpc/minrpc_server.h:181]: also, s/char/uint8_t/?

File: src/runtime/rpc/minrpc/minrpc_server.h [hunk 23]
```diff
@@ -0,0 +1,598 @@
...
+    this->Read(&ctx);
+    this->Read(&type_hint);
+    int call_ecode = 0;
+
+    if (ctx.device_type == kDLCPU) {
+       char* dptr = reinterpret_cast<char*>(handle) + offset;
+       this->ReadArray(dptr, num_bytes);
+    } else {
+      char* temp_data = this->ArenaAlloc<char>(num_bytes);
+      this->ReadArray(temp_data, num_bytes);
```
Comment (areusch) [inline src/runtime/rpc/minrpc/minrpc_server.h:218]: for the micro runtime, we will need to avoid this copy step, here and in CopyFromRemote. is there a corresponding change we need to make to TVMDevice API to facilitate that?

File: src/runtime/rpc/rpc_endpoint.cc [hunk 24]
```diff
@@ -0,0 +1,1059 @@
...
+  RPCCode HandleNextEvent(bool client_mode, RPCSession::FEncodeReturn setreturn) {
+    std::swap(client_mode_, client_mode);
+
+    while (this->Ready()) {
+      switch (state_) {
+        case kInitHeader: HandleInitHeader(); break;
+        case kRecvPacketNumBytes: {
+          uint64_t packet_nbytes;
+          CHECK(this->Read(&packet_nbytes));
+          if (packet_nbytes != 0) {
```
Comment (areusch) [inline src/runtime/rpc/rpc_endpoint.cc:124]: check against a max?

File: jvm/core/src/main/java/org/apache/tvm/contrib/GraphRuntime.java (hunk 3 above)
Comment (tqchen) [reply jvm/core/src/main/java/org/apache/tvm/contrib/GraphRuntime.java:42]: give it is in the java runtime and not in the scope of this PR, will leave as it is. There is also a debatable tradeoff (common usage vs less abstraction)

File: src/runtime/rpc/rpc_pipe_impl.cc [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+  int parent_write = parent2child[1];
+  int child_read = parent2child[0];
+  int child_write = child2parent[1];
+
+  pid_t pid = fork();
+  if (pid == 0) {
+    // child process
+    close(parent_read);
+    close(parent_write);
+    std::string sread_pipe = std::to_string(child_read);
```
Comment (areusch) [inline src/runtime/rpc/rpc_pipe_impl.cc:98]: can we use dup2 here instead, or is that not portable?

File: python/tvm/contrib/cc.py (hunk 7 above)
Comment (tqchen) [reply python/tvm/contrib/cc.py:94]: I think it is mainly a difference in API:

- A0  pushing more flags to export_library 
- A1 using functional style programming decorator, to decorate a existing compiler by adding ldflags implicitly. Pass in fcompile variants, or use fcompile->fcompile functors to decorate new compilers.

My take is that A1 is as powerful as A0, and the complexity of compiler customization goes to the decoration process.

We can certainly de-couple the APIs in A1. Right now compilation and linking are coupled, and a more powerful way to construct A1 might be as follows, and we can work toward that direction. 
```python
fcompile = attach_linker_opts(get_compiler())
```

File: src/runtime/rpc/minrpc/posix_popen_server.cc [hunk 26]
```diff
@@ -0,0 +1,74 @@
...
+#include <cstdlib>
+#include "minrpc_server.h"
+
+namespace tvm {
+namespace runtime {
+
+/*!
+ * \brief IOHandler based on posix API.
+ */
+class PosixIOHandler {
```
Comment (areusch) [inline src/runtime/rpc/minrpc/posix_popen_server.cc:33]: will there also be a flush? it would be helpful in implementing simpler e.g. ethernet, usb, etc transports.

File: python/tvm/rpc/client.py (hunk 11 above)
Comment (tqchen) [reply python/tvm/rpc/client.py:216]: good pt about existence check. abspath makes the execution more consistent, e.g. if we need to chpwd in the subsequent functions

File: python/tvm/rpc/server.py (hunk 16 above)
Comment (areusch) [reply python/tvm/rpc/server.py:60]: okay, just documenting that I found this kwarg confusing when first learning the code :)

File: python/tvm/rpc/minrpc.py (hunk 14 above)
Comment (tqchen) [reply python/tvm/rpc/minrpc.py:37]: This is temporary hack. Ideally we should ship min_server to a location (e.g. build) that get installed when we do `setup.py install` instead of relying on the availablity of the source. Or we just ship the source with the installation.

When we figure that out(possibly in micro tvm link), we can have a consolidated  library as in the libinfo.find_library()

File: src/runtime/rpc/minrpc/minrpc_server.h (hunk 21 above)
Comment (tqchen) [reply src/runtime/rpc/minrpc/minrpc_server.h:178]: The same code can be used to host a TVM runtime with GPU support

File: src/runtime/rpc/minrpc/minrpc_server.h (hunk 20 above)
Comment (tqchen) [reply src/runtime/rpc/minrpc/minrpc_server.h:84]: right now there is no maximum len limit

File: python/tvm/contrib/cc.py (hunk 7 above)
Comment (areusch) [reply python/tvm/contrib/cc.py:94]: okay, I think de-coupling the APIs in A1 makes sense. if we do that though, then only autotvm (and other user scripts and any µTVM veneers) need to care about linker flags and linker scripts, correct? we can remove that complexity from `export_library`.

File: python/tvm/rpc/client.py (hunk 11 above)
Comment (areusch) [reply python/tvm/rpc/client.py:216]: I guess i'm wondering why we would need to do that, and wondering if it's more motivated by your rpc server test?

File: src/runtime/rpc/minrpc/minrpc_server.h (hunk 22 above)
Comment (tqchen) [reply src/runtime/rpc/minrpc/minrpc_server.h:181]: Arena itself will call allocator's allocate, which will call IOHandler.ThrowError

File: src/runtime/rpc/rpc_pipe_impl.cc (hunk 25 above)
Comment (tqchen) [reply src/runtime/rpc/rpc_pipe_impl.cc:98]: do you mean dup2 to stdio? My only worry is that what if the runtime do something like print to stdout. For now most of the runtime only logs to stderr so we might be fine. But passing in pipe number might be more robust in this case

File: src/runtime/rpc/minrpc/posix_popen_server.cc (hunk 26 above)
Comment (tqchen) [reply src/runtime/rpc/minrpc/posix_popen_server.cc:33]: Good point, perhaps we can add that as a followup.

File: src/runtime/rpc/minrpc/minrpc_server.h (hunk 23 above)
Comment (tqchen) [reply src/runtime/rpc/minrpc/minrpc_server.h:218]: It only happens in the non-CPU case(which need the copy).

File: src/runtime/rpc/rpc_endpoint.cc (hunk 24 above)
Comment (tqchen) [reply src/runtime/rpc/rpc_endpoint.cc:124]: there is no max limit for now.

File: src/runtime/rpc/minrpc/minrpc_server.h (hunk 23 above)
Comment (areusch) [reply src/runtime/rpc/minrpc/minrpc_server.h:218]: so for micro would we similarly override kDLMicroDev

Comment (tqchen) [reply src/runtime/rpc/minrpc/minrpc_server.h:218]: hmm, on device microTVM rpc server, I believe it is better to set the device code to directly be kDLCPU(as it is CPU from the pt of view of the microcontroller).

If the code is kDLMicroDev, that means the current machine is a host machine that drives the micro tvm, and the temp memory is on the host

File: python/tvm/rpc/client.py (hunk 11 above)
Comment (tqchen) [reply python/tvm/rpc/client.py:216]: You are right that there is none atm, we could certainly skip this part.

File: python/tvm/contrib/cc.py (hunk 7 above)
Comment (tqchen) [reply python/tvm/contrib/cc.py:94]: I think your understanding is correct

File: python/tvm/_ffi/base.py (hunk 6 above)
Comment (tqchen) [reply python/tvm/_ffi/base.py:54]: Note that the change to LD_LIBRARY_PATH won't affert the execution after the python script exits.   I still think it makes sense to make the libary path in the installation to be available for sub-processes

Comment (areusch) [reply python/tvm/_ffi/base.py:54]: correct me if i'm wrong, doesn't this mean that any python script that imports tvm and then launches a subprocess will launch a subprocess with LD_LIBRARY_PATH set? we really don't want that. if the subprocess depends on any library that happens to be in the same dir as libtvm.so, it will wind up loading that instead of the system-installed copy. this is a really subtle bug that is very difficult to track down.

File: src/runtime/rpc/rpc_pipe_impl.cc (hunk 25 above)
Comment (areusch) [reply src/runtime/rpc/rpc_pipe_impl.cc:98]: oh okay, that makes sense to me. I think this is fine then.

File: .gitignore [hunk 27]
```diff
@@ -2,9 +2,10 @@
 __pycache__/
 *.py[cod]
 *$py.class
-
+*.S
```
Comment (areusch) [inline .gitignore:5]: why this change?

**Comment by areusch** [comment]
I think this looks good to me now!

File: .gitignore (hunk 27 above)
Comment (tqchen) [reply .gitignore:5]: try to ignore assembly files generated in commands.

**Comment by tqchen** [comment]
@tmoreau89 @areusch  Thanks, please https://tvm.apache.org/docs/contribute/code_review.html?highlight=approve#approve-and-request-changes-explicitly

File: .gitignore (hunk 27 above)
Comment (areusch) [reply .gitignore:5]: ah ok. I guess this means we can't check in assembly without removing it, but we can do that later too.
